pkg/registries/registries.go

```diff
@@ -4,20 +4,32 @@ import (
 	"sort"
 	"strings"
 
-	"github.com/containers/image/pkg/sysregistriesv2"
+	"github.com/containers/image/v5/pkg/sysregistriesv2"
 	apioperatorsv1alpha1 "github.com/openshift/api/operator/v1alpha1"
 )
 
-// scopeMatchesRegistry returns true if a scope value (as in sysregistriesv2.Registry.Prefix / sysregistriesv2.Endpoint.Location)
-// matches a host[:port] value in reg.
-func scopeMatchesRegistry(scope, reg string) bool {
+// regMatchesScope returns true if a scope value (as in sysregistriesv2.Registry.Prefix / sysregistriesv2.Endpoint.Location)
+// matches a host[:port] value in reg or if it is a sub-scope of reg.
+func regMatchesScope(scope, reg string) bool {
+	match := false
 	if reg == scope {
 		return true
 	}
+	// return true if a scope value matches a host[:port] value in reg
 	if len(scope) > len(reg) {
-		return strings.HasPrefix(scope, reg) && scope[len(reg)] == '/'
+		match = strings.HasPrefix(scope, reg) && scope[len(reg)] == '/'
 	}
-	return false
+	// return true if scope is a value that is a sub-scope of reg
+	// e.g *.foo.example.com is a sub-scope of *.example.com or bar.example.com/bar is a sub-scope of *.example.com
+	if strings.HasPrefix(reg, "*.") {
+		match = strings.HasSuffix(scope, reg[1:]) || strings.Contains(scope, reg[1:]+"/") || strings.Contains(scope, reg[1:]+":")
+		// Check that we are not matching on namespace or repo e.g *.foo should not match quay/bar.foo or quay/bar.foo/example or quay/bar.foo:400
+		regIndex := strings.Index(scope, reg[1:])
+		if regIndex > 0 && strings.Contains(scope[:regIndex+1], "/") {
```

Comment:
This looks convoluted. Why first match with an inaccurate `contains` on the full string, and then see if we found a `/`? It seems to me that we could extract the part of `reg` that precedes a `/`, if any (and `:`), and then do a simpler HasSuffix match.

OTOH if it works…

---

`regIndex < 0` looks like an impossible situation.

Reply:
If there is no index match for the substring, it returns a -1. That is why I have the regIndex > 0 check.

Reply:
And fixed to suggested logic